web/api/backend.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+	"time"
+
+	jwtmiddleware "github.com/auth0/go-jwt-middleware"
+	"github.com/dgrijalva/jwt-go"
+
+	"github.com/gorilla/mux"
+)
+
+// Claims object that will be encoded to a JWT.
+// We add jwt.StandardClaims as an embedded type, to provide fields like expiry time.
+type Claims struct {
+	UploadCredentials map[string]interface{}
+	jwt.StandardClaims
+}
+
+var jwtKey = []byte("my_secret_key")
+var creds WrapperCommand
+
+// CredentialsHandler takes as input a json body with the parts of the command
+// to be executed and a json body representing the uploaded credentials.json file and returns
+// a created jwt token that will be sent to the front end and cached for reused if needed
+// returns a 401 status if jwt token cannot be created.
+func CredentialsHandler(w http.ResponseWriter, r *http.Request) {
+	setupResponse(&w, r)
+	if (*&r).Method == "OPTIONS" {
+		return
+	}
+
+	// Get the JSON body and decode into credentials.
+	err := json.NewDecoder(r.Body).Decode(&creds)
+	if err != nil {
+		// If the structure of the body is wrong, return an HTTP error.
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if len(creds.Credential) == 0 {
+		w.WriteHeader(http.StatusBadRequest)
+		io.WriteString(w, `{"error":"cannot make token without credentials"}`)
+		return
+	}
+
+	// Declare the expiration time of the token
+	// here, we have kept it as 1 day.
+	expirationTime := time.Now().Add(1440 * time.Minute)
+	// Create the JWT claims, which includes the uploaded credentials json body and expiry time.
+	claims := &Claims{
+		UploadCredentials: creds.Credential,
+		StandardClaims: jwt.StandardClaims{
+			// In JWT, the expiry time is expressed as unix milliseconds.
+			ExpiresAt: expirationTime.Unix(),
+		},
+	}
+
+	// Declare the token with the algorithm used for signing, and the claims.
+	signedCredentials := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	// Create the JWT string
+	credentialsString, err := signedCredentials.SignedString(jwtKey)
+	if err != nil {
+		// If there is an error in creating the JWT return an internal server error.
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	io.WriteString(w, `{"token":"`+credentialsString+`"}`)
+
+}
+
+// AuthHandler checks if the token specified in the authorization bearer of the http request
+// is valid, or not expired and created by the web server.
```

Comment:
'not expired' can be dropped since it is included in invalid.

Reply:
Dropped in new version.